skorch/tests/test_hf.py

```diff
@@ -755,3 +757,376 @@ def train_step(self, *args, **kwargs):
+        )
+        checkpoint = TrainEndCheckpoint(
+            f_pickle=storer,
+            f_params=None,
+            f_optimizer=None,
+            f_criterion=None,
+            f_history=None,
+        )
+        net.set_params(callbacks=[checkpoint])
+        net.fit(*data)
+
+        assert len(mock_hf_api.calls) == storer._call_count == 1
+        obj, _ = mock_hf_api.calls[0]
+        assert isinstance(obj, io.IOBase)
+
+    def test_train_end_checkpoint_torch_save(
+            self, net, data, mock_hf_api, hf_hub_storer_cls
+    ):
+        # f_pickle uses pickle but f_params et al use torch.save, which works a
+        # bit differently. Therefore, we need to test both.
+        from skorch.callbacks import TrainEndCheckpoint
+
+        storer = hf_hub_storer_cls(
+            mock_hf_api,
+            path_in_repo='weights.pt',
+            repo_id='my-user/my-repo',
+            buffered=True,
+        )
+        checkpoint = TrainEndCheckpoint(
+            f_params=storer,
+            f_optimizer=None,
+            f_criterion=None,
+            f_history=None,
+        )
+        net.set_params(callbacks=[checkpoint])
+        net.fit(*data)
+
+        assert len(mock_hf_api.calls) == storer._call_count == 1
+        obj, _ = mock_hf_api.calls[0]
+        assert isinstance(obj, io.IOBase)
+
+    def test_checkpoint_pickle(self, net, data, mock_hf_api, hf_hub_storer_cls):
+        # Checkpoint saves the model multiple times
+        from skorch.callbacks import Checkpoint
+
+        storer = hf_hub_storer_cls(
+            mock_hf_api, path_in_repo='my-model', repo_id='my-user/my-repo', token='123'
+        )
+
+        checkpoint = Checkpoint(
+            f_pickle=storer,
+            f_params=None,
+            f_optimizer=None,
+            f_criterion=None,
+            f_history=None,
+        )
+
+        net.set_params(callbacks=[checkpoint], max_epochs=10)
+        net.fit(*data)
+
+        # each time the valid loss improves, there should be a checkpoint
+        num_checkpoints_expected = sum(net.history[:, 'valid_loss_best'])
+        num_checkpoints_actual = len(mock_hf_api.calls)
+        assert num_checkpoints_actual == num_checkpoints_expected
+
+    def test_checkpoint_torch_save(self, net, data, mock_hf_api, hf_hub_storer_cls):
+        from skorch.callbacks import Checkpoint
+
+        storer = hf_hub_storer_cls(
+            mock_hf_api, path_in_repo='my-model', repo_id='my-user/my-repo', token='123'
+        )
+
+        checkpoint = Checkpoint(
+            f_params=None,
+            f_optimizer=storer,
+            f_criterion=None,
+            f_history=None,
+        )
+
+        net.set_params(callbacks=[checkpoint], max_epochs=10)
+        net.fit(*data)
+
+        # each time the valid loss improves, there should be a checkpoint
+        num_checkpoints_expected = sum(net.history[:, 'valid_loss_best'])
+        num_checkpoints_actual = len(mock_hf_api.calls)
```

Comment:
this test should make sure that the number is > 0. while it is valid that there might not be best losses (currently there should be at least one but that might change), the test would be useless if #best = 0.